### PR TITLE
Edit Decorations

### PR DESCRIPTION
**Pull request by chevreaux**
Decorations on the preview and in the sidebar. Sprites ready to go once this is in.

Essentially, any component that works with objects can now take and object or a deco. (I'm using 'Entity' to refer to objects and decos together.) I tried to do this by having a single prop with a union type (ex: Entity: GameObjectType | DecorationType). That would require a lot more work though. Instead:

- LevelPreviewEntities, SidebarEntityAdder and WorldDataContext reducer actions take in a string that tells them which type they are for.
- SidebarEntityList and SidebarEntityItem have a deco? and an obj? prop, and only receive one from their parent.

There is more code repetition than I would like, but I can't think of a better way. 

Probably the biggest issue right now is the Select menu for adding decorations. It's really laggy with 1700 entries.

### CODE REVIEW

File: src/levelEditor/sidebar/objectsList/SidebarObjectText.js [hunk 1]
```diff
@@ -6,12 +6,14 @@ import type {GameObjectType} from '../../types/GameObjectType';
 import convertLevelIdToCoordinates from '../../util/convertLevelIdToCoordinates';
 
 type Props = $ReadOnly<{
-	obj: GameObjectType,
+	obj: ?GameObjectType,
```
Comment (ngyikp) [inline src/levelEditor/sidebar/objectsList/SidebarObjectText.js:9]: Not needed if SidebarEntityItem.js checks that the object is non-null
```suggestion
	obj: GameObjectType,
```

File: src/levelEditor/types/EntityCategories.js [hunk 2]
```diff
@@ -0,0 +1,2119 @@
...
+		'objProp_placed',
+		'objPropspritelayer',
+		'objRider',
+		'objSnakeburst',
+		'objSongNode',
+		'objTextfield',
+		'objWordcount',
+	];
+
+export const SPRITES: $ReadOnlyArray<SpriteType> = [
```
Comment (ngyikp) [inline src/levelEditor/types/EntityCategories.js:417]: Split the sprites/game objects to their own files, at least to not ruin the Git blame history

File: src/levelEditor/sidebar/properties/PropertyNumberInput.js [hunk 3]
```diff
@@ -6,12 +6,14 @@ import styles from './PropertyNumberInput.module.css';
...
 	onCommitValue: (newValue: number | string) => mixed,
-	initialValue: number | string, // it could a number, string of number, or empty string
+	initialValue: number | string, // it could a number, string of number, or empty string,
+	step?: number,
 }>;
 
 export default function PropertyNumberInput({
 	onCommitValue,
 	initialValue,
+	step,
```
Comment (ngyikp) [inline src/levelEditor/sidebar/properties/PropertyNumberInput.js:16]:
```suggestion
	step = 1,
```

File: src/levelEditor/types/PlaceableType.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+// @flow strict
+
+import type {GameObjectEntityType} from './GameObjectEntityType';
+import type {SpriteType} from './SpriteType';
+
+export type PlaceableType = {
+	type: 'OBJECT' | 'DECO',
+	data: GameObjectEntityType | SpriteType,
+};
```
Comment (ngyikp) [inline src/levelEditor/types/PlaceableType.js:9]: Does this not work?
```suggestion
export type PlaceableType = {
	type: 'OBJECT',
	data: GameObjectEntityType,
} | {
	type: 'DECO',
	data: SpriteType,
};
```

**Review by ngyikp — CHANGES_REQUESTED** [review]

File: src/levelEditor/preview/LevelPreview.js [hunk 5]
```diff
@@ -52,17 +55,16 @@ export default function LevelPreview(props: Props): React$Node {
...
-					onObjectClick={props.onObjectClick}
-					onObjectHover={props.onObjectHover}
+					entityIndexHover={props.objectIndexHover}
+					onEntityClick={props.onEntityClick}
+					onEntityHover={props.onObjectHover}
+					type="OBJECT"
 				/>
 			) : null}
 
-			{props.activeUiViews.includes('OBJECTS') &&
```
[the remark this answers is not in the document]
Comment (ngyikp) [reply src/levelEditor/preview/LevelPreview.js:64]: I fixed it in [62f418a](https://github.com/chicory-pizza/chicory-data/commit/62f418a3df8ef8d47f0e09945f318bb881edbb1d)

File: .prettierignore [hunk 6]
```diff
@@ -7,3 +7,4 @@ package-lock.json
 
 src/levelEditor/level_data.json
 src/levelEditor/__mocks__/level_data.json
+src/levelEditor/spriteData.json
```
[the remark this answers is not in the document]
Comment (chevreaux) [reply .prettierignore:10]: I've written a script for UndertaleModTool that extracts the data for all sprites.

File: src/levelEditor/types/PlaceableType.js (hunk 4 above)
Comment (chevreaux) [reply src/levelEditor/types/PlaceableType.js:9]: This causes SidebarEntityAdder to complain in a way I don't know how to fix.

File: src/levelEditor/preview/LevelPreviewEntities.module.css [hunk 7]
```diff
@@ -3,6 +3,10 @@
 	z-index: 3;
 }
 
-.itemHover {
+.objHover {
 	background: rgb(216, 255, 85);
 }
+
+.decoHover {
+	background: rgb(85, 96, 255);
```
[the remark this answers is not in the document]
Comment (chevreaux) [reply src/levelEditor/preview/LevelPreviewEntities.module.css:11]: I'll use Luncheon purple instead.

File: src/levelEditor/preview/LevelPreviewEntities.js [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+type Props = $ReadOnly<{
+	level: LevelType,
+	entityIndexHover: ?number,
+	onEntityClick: (entityIndex: number, entityType: GameEntityType) => mixed,
+	onEntityHover: (entityIndex: ?number) => mixed,
+	type: GameEntityType,
+}>;
+
+function LevelPreviewEntities(props: Props): React$Node {
+	if (props.type === 'OBJECT') {
```
[the remark this answers is not in the document]
Comment (chevreaux) [reply src/levelEditor/preview/LevelPreviewEntities.js:21]: Hah, I had something like this in an earlier version. I can make that change some time later.

File: src/levelEditor/types/PlaceableType.js (hunk 4 above)
Comment (ngyikp) [reply src/levelEditor/types/PlaceableType.js:9]: Yeah: https://flow.org/en/docs/types/unions/#toc-disjoint-object-unions

File: src/levelEditor/sidebar/objectsList/SidebarObjectText.js [hunk 9]
```diff
@@ -11,7 +11,9 @@ type Props = $ReadOnly<{
 
 export default function SidebarObjectText({obj}: Props): React$Node {
 	const [, setNewCoordinates] = useCurrentCoordinatesNonNullable();
-
+	if (obj == null) {
+		return 'Null';
+	}
```
Comment (ngyikp) [inline src/levelEditor/sidebar/objectsList/SidebarObjectText.js:16]: obj is never non-null, this is unneeded
```suggestion
```

File: .prettierignore (hunk 6 above)
[the remark this answers is not in the document]
Comment (ngyikp) [reply .prettierignore:10]: Can you share the script somewhere such as a repo or gist?

**Review by ngyikp — APPROVED** [review]
LGTM, basic functionality works well, let's fix any new issues on a new PR

**Pull request merged**